src/core/global.c

```diff
@@ -303,20 +307,28 @@ static void nn_global_term (void)
 void nn_term (void)
 {
     int i;
+    struct nn_sock *sock;
 
     nn_glock_lock ();
 
     /*  Switch the global state into the zombie state. */
     self.flags |= NN_CTX_FLAG_ZOMBIE;
 
     /*  Mark all open sockets as terminating. */
-    if (self.socks && self.nsocks) {
+    if (self.nsocks != 0) {
         for (i = 0; i != NN_MAX_SOCKETS; ++i)
-            if (self.socks [i])
+            if (self.socks [i] != NULL) {
                 nn_sock_zombify (self.socks [i]);
+            }
     }
 
     nn_glock_unlock ();
+
+    /* Make sure we really close resources, this will cause global
+       resources to be freed too when the last socket is closed. */
+    for (i = 0; i < NN_MAX_SOCKETS; i++) {
+        (void) nn_close_impl (i, 1);
+    }
```

Comment:
PS -- I understand that if the glock needs to be held around the loop, then `nn_close_impl` would need to be refactored (to not glock), which would cascade these same considerations anywhere `nn_close_impl` is called.